backends/graphics/gles-custom/gles-custom.cpp

```diff
@@ -0,0 +1,703 @@
+	*waiting_for_flip = 0;
+}
+
+static drmEventContext evctx;
+
+void drm_fb_destroy_callback(struct gbm_bo *bob, void *data);
+
+struct drm_fb * drm_fb_get_from_bo(struct gbm_bo *bob)
+{
+	struct drm_fb *fbu = (drm_fb*)gbm_bo_get_user_data(bob);
+	uint32_t width, height, stride, handle;
+
+	if (fbu) {
+		return fbu;
+	}
+	
+	fbu = (drm_fb*)calloc(1, sizeof *fbu);
+	fbu->bo = bob;
+
+	width = gbm_bo_get_width(bob);
+	height = gbm_bo_get_height(bob);
+	stride = gbm_bo_get_stride(bob);
+	handle = gbm_bo_get_handle(bob).u32;
+
+	if (drmModeAddFB(drm.fd, width, height, 24, 32, stride, handle, &fbu->fb_id)) {
+		debug("Could not add drm framebuffer\n");
+		free(fbu);
+		return NULL;
+	}
+
+	// We used to pass the destroy callback function here. Now it's done manually in deinit_egl()
+	gbm_bo_set_user_data(bob, fbu, NULL);
+	return fbu;
+}
+
+void DRM_PageFlip(void){
+	struct gbm_bo *next_bo;
+	int waiting_for_flip = 1;
+
+	next_bo = gbm_surface_lock_front_buffer(gbm.surface);
+	fb = drm_fb_get_from_bo(next_bo);
+
+	if (drmModePageFlip(drm.fd, drm.crtc_id, fb->fb_id,
+			DRM_MODE_PAGE_FLIP_EVENT, &waiting_for_flip)){
+		debug ("Failed to queue pageflip\n");
+		return;
+	}
+
+	while (waiting_for_flip) {
+		FD_ZERO(&fds);
+		
+		FD_SET(0, &fds);
+		FD_SET(drm.fd, &fds);
+
+		select(drm.fd+1, &fds, NULL, NULL, NULL);
+		
+		drmHandleEvent(drm.fd, &evctx);
+	}
+	
+	// release last buffer to render on again
+	gbm_surface_release_buffer(gbm.surface, bo);
+	bo = next_bo;
+}
+
+int init_drm(void)
+{
+	static const char *modules[] = {
+			"i915", "radeon", "nouveau", "vmwgfx", "omapdrm", "exynos", "msm"
+	};
+	drmModeRes *resources;
+	drmModeConnector *connector = NULL;
+	drmModeEncoder *encoder = NULL;
+	unsigned i, area;
+
+	// In plain C, we can just init evctx at declare time, but it's now allowed in C++
+	evctx.version = DRM_EVENT_CONTEXT_VERSION;
+	evctx.page_flip_handler = page_flip_handler;
+
+	for (i = 0; i < (sizeof(modules) / sizeof(modules[0])); i++) {
+		debug("trying to load module %s...", modules[i]);
+		drm.fd = drmOpen(modules[i], NULL);
+		if (drm.fd < 0) {
+			debug("failed.");
+		} else {
+			debug("success.");
+			break;
+		}
+	}
+
+	if (drm.fd < 0) {
+		debug("could not open drm device\n");
+		return -1;
+	}
+
+	resources = drmModeGetResources(drm.fd);
+	if (!resources) {
+		debug("drmModeGetResources failed\n");
+		return -1;
+	}
+
+	// find a connected connector
+	for (i = 0; i < (unsigned)resources->count_connectors; i++) {
+		connector = drmModeGetConnector(drm.fd, resources->connectors[i]);
+		if (connector->connection == DRM_MODE_CONNECTED) {
+			// it's connected, let's use this!
+			break;
+		}
+		drmModeFreeConnector(connector);
+		connector = NULL;
+	}
+
+	if (!connector) {
+		// we could be fancy and listen for hotplug events and wait for
+		// a connector..
+		debug("no connected connector found\n");
+		return -1;
+	}
+	// find highest resolution mode
+	for (i = 0, area = 0; i < (unsigned)connector->count_modes; i++) {
```

Comment:
C style once again. Why declare variable at the beginning of the function? and 'unsigned' what? you should be explicit there for readability